### PR TITLE
Issue 23895 creaste api method to get result

### PR DESCRIPTION
**Pull request by freddyucv**
### Proposed Changes
* Fix misspelling in referer REACH PAGE Metric parameter
https://github.com/dotCMS/core/pull/23973/files#diff-039db8b104840e0437baac989d496b376f2d496bc9fc89bb79046e38e9d563d3R574

https://github.com/dotCMS/core/pull/23973/files#diff-a99d90c3cfd57cac21364804699421782fdb91332fccc3b56ad0441b375b7ba4R19

* Turn AnalyticsHelper into a Singleton and all the methods to no static, I need this class to be able to be mock 

https://github.com/dotCMS/core/pull/23973/files#diff-40a3f5b2dc2a91b5dcd987cc6beadd58c5f4e6077e680f58bd995b1e6ae971deR38

This force me to change several clases to use the AnalyticsHelper as a Singleton like:

https://github.com/dotCMS/core/pull/23973/files

* Allow to set a Goal in the ExperimentDataGen

https://github.com/dotCMS/core/pull/23973/files#diff-e2bfadee0656f6eb9dd29d02c0cfe90c1c4d23f540683e6326a8a770e50646aaR93

before this changes the Goal was hard coding all the time

https://github.com/dotCMS/core/pull/23973/files#diff-e2bfadee0656f6eb9dd29d02c0cfe90c1c4d23f540683e6326a8a770e50646aaL76

* Change the return method here

https://github.com/dotCMS/core/pull/23973/files#diff-29cacbde2fe6ea160809f524ffd80ffc8526f8d2f60cff0339996965c0e36b90R50

I need to mock the return object 

* Create a Function to apply the Operator for two values

https://github.com/dotCMS/core/pull/23973/files#diff-75510bc7dededdfed8b986435f8be1f23e53e726a1f3e3e79e9b2b48e8930cd6R35

This allow me to check if the condition if valid for a specific Event into the Experiment. it is used here

https://github.com/dotCMS/core/pull/23973/files#diff-10fc5a5c8d8484aa3f69421a60d72d73aebe0a65413ce9023843ddbf7f9a38c7R49

* Create EventType concept now with this clase we can filter o do whatever We want using the enum class and not literal String

https://github.com/dotCMS/core/pull/23973/files#diff-2c88496ab59f4af1dc7b5faaca4017596107ba5909b0e996b06478f6969a6a97R5

used here:

https://github.com/dotCMS/core/pull/23973/files#diff-2e8eed231d46c15f3a24986a33f0800b01d807cec95d5b98343efaa0b8d98d45R42

* We need a merge method into CubeJSQuery to merge two partial CubeJSQuery 

https://github.com/dotCMS/core/pull/23973/files#diff-c50354f1a06c2e8525d71ddbb42ebb62b376ba7a22d238ff714e76584d3c3674R194

this method is used here 

https://github.com/dotCMS/core/pull/23973/files#diff-86803e91e3ce9f9735e92892ac7725c5a865a653a7db53abe5d4e928248defc2R101

* Create a method into the Experiment API to get all the Events from a specific Experiment

https://github.com/dotCMS/core/pull/23973/files#diff-8bf1865184fd59c5c64882df72bd0c71a403a544777830acf250bbf54ea84d90R167

* Create a method to calculate the partial or total result of a Experiment

https://github.com/dotCMS/core/pull/23973/files#diff-8bf1865184fd59c5c64882df72bd0c71a403a544777830acf250bbf54ea84d90R160

* Create a Util classes to analyze a set of events to realized how many times the Goal was success into a Exeriment

https://github.com/dotCMS/core/pull/23973/files#diff-1fdf3602f53e848245929cb77e96e83012c452faf87ae01d3e699826cfd22f2eR25

https://github.com/dotCMS/core/pull/23973/files#diff-b75692fe81272019be341546b06800bd71b258fe7c302b05512aa2f3ba7b1141R7

https://github.com/dotCMS/core/pull/23973/files#diff-10fc5a5c8d8484aa3f69421a60d72d73aebe0a65413ce9023843ddbf7f9a38c7R20

* Create util clases to build a CubeJS Query to get all the events from a Experiments

https://github.com/dotCMS/core/pull/23973/files#diff-86803e91e3ce9f9735e92892ac7725c5a865a653a7db53abe5d4e928248defc2R59

https://github.com/dotCMS/core/pull/23973/files#diff-703956d3a3cfc95fe8aa61b89afce9162b12d26f8f8b53f22eff019eb9b07873R11

https://github.com/dotCMS/core/pull/23973/files#diff-703956d3a3cfc95fe8aa61b89afce9162b12d26f8f8b53f22eff019eb9b07873R11

* Create a way into the HttpMockServer to validate if a URL was hit during the test

https://github.com/dotCMS/core/pull/23973/files#diff-bb8e6d4042484875cf32b80f746a82828bf3c93ebaa404c732329defc4f9b003R115

https://github.com/dotCMS/core/pull/23973/files#diff-de5f0272efb882c609696065290263f8dbfa7d10128686f601eaab3168077b31R142

* Fixing a master broken test
https://github.com/dotCMS/core/pull/23973/files#diff-07bf99cd1d156dc4ea1bab27b0612fa51e3fa3cb2b52c5c4153366bfb416cc51R118

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/analytics/metrics/AbstractCondition.java [hunk 1]
```diff
@@ -23,8 +23,30 @@ public interface AbstractCondition {
     String value();
 
     enum Operator {
-        EQUALS,
-        CONTAINS
+        EQUALS((value1, value2) -> value1.equals(value2)),
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/analytics/metrics/AbstractCondition.java:26]: Maybe this is not a good approach to the EQUALS operator when we are comparing URL, maybe in the URL case we need to not take account the query params or support relative path to not include the domain name or protocol in the comparation.

File: dotCMS/src/main/java/com/dotcms/analytics/metrics/AbstractCondition.java [hunk 2]
```diff
@@ -23,8 +23,22 @@ public interface AbstractCondition {
...
+        CONTAINS((value1, value2) -> value1.toString().contains(value2.toString()));
+
+        private OperatorFunc function;
+
+        Operator(final OperatorFunc func){
+            this.function = func;
+        }
+
+        public OperatorFunc getFunction() {
```
[the remark this answers is not in the document]
Comment (jcastro-dotcms) [reply dotCMS/src/main/java/com/dotcms/analytics/metrics/AbstractCondition.java:39]: What about `ComparisonFunction` ?

**Review by jcastro-dotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/analytics/helper/AnalyticsHelper.java [hunk 3]
```diff
@@ -34,13 +35,21 @@
  */
 public class AnalyticsHelper {
 
+    private static Lazy<AnalyticsHelper> analyticsHelper = Lazy.of(() -> new AnalyticsHelper());
```
Comment (victoralfaro-dotcms) [inline dotCMS/src/main/java/com/dotcms/analytics/helper/AnalyticsHelper.java:38]: Not sure what is the benefit of this.

File: dotCMS/src/main/java/com/dotcms/analytics/metrics/MetricType.java [hunk 4]
```diff
@@ -16,7 +16,7 @@ public enum MetricType {
     REACH_PAGE(new Builder()
             .goalName("Maximize Reaching a Page")
             .allRequiredParameters (Parameter.builder().name("url").build()) //TODO we can create singletons of these Parameters in order to reuse
-            .optionalParameters(Parameter.builder().name("referrer").build())
+            .optionalParameters(Parameter.builder().name("referer").build())
```
Comment (victoralfaro-dotcms) [inline dotCMS/src/main/java/com/dotcms/analytics/metrics/MetricType.java:19]: This is a typo. Referrer is correct.

File: dotCMS/src/main/java/com/dotcms/analytics/helper/AnalyticsHelper.java (hunk 3 above)
Comment (victoralfaro-dotcms) [reply dotCMS/src/main/java/com/dotcms/analytics/helper/AnalyticsHelper.java:38]: nvm

**Review by victoralfaro-dotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/experiments/business/ExperimentAPIImpIT.java [hunk 5]
```diff
@@ -204,4 +244,618 @@ public void testADeleteOriginalVariant_shouldFail() throws DotDataException, Dot
         APILocator.getExperimentsAPI().deleteVariant(newExperiment.id().orElse(""), originalVariant.id(),
                 APILocator.systemUser());
     }
+
+    /**
+     * Method to test: {@link ExperimentsAPIImpl#getEvents(Experiment)}
+     * When: Call the methods with a Experiment with 4 Session each one with several
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/experiments/business/ExperimentAPIImpIT.java:250]: Don't understand this text

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPI.java [hunk 6]
```diff
@@ -149,4 +150,19 @@ Optional<Rule> getRule(final Experiment experiment)
...
      */
     boolean isAnyExperimentRunning() throws DotDataException;
+
+    /**
+     * Return the Experiment partial or total result.
+     *
+     * @param experiment
+     * @return
+     */
+    ExperimentResult getResult(final Experiment experiment);
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPI.java:160]: Rename to `getResults` and the object to `ExperimentResults`

File: dotCMS/src/main/java/com/dotcms/experiments/business/result/ExperimentResultQueryFactory.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+ * experiment_identifier: is the {@link Experiment}'s id.
+ *
+ * - Dynamic part: this part depends of the {@link com.dotcms.experiments.model.Goals} of the
+ * {@link Experiment}.
+ *
+ * both part are merged using {@link CubeJSQuery.Builder#merge(CubeJSQuery, CubeJSQuery)} method.
+ *
+ * @see MetricExperimentResultQuery
+ */
+public enum ExperimentResultQueryFactory {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/experiments/business/result/ExperimentResultQueryFactory.java:61]: ExperimentResultsQueryFactory

File: dotCMS/src/main/java/com/dotcms/experiments/business/result/MetricExperimentAnalyzer.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package com.dotcms.experiments.business.result;
+
+import com.dotcms.analytics.metrics.Metric;
+import com.dotcms.cube.CubeJSResultSet;
+import java.util.List;
+
+public interface MetricExperimentAnalyzer {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/experiments/business/result/MetricExperimentAnalyzer.java:7]: Doc

File: dotCMS/src/integration-test/java/com/dotcms/experiments/business/ExperimentAPIImpIT.java (hunk 5 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/experiments/business/ExperimentAPIImpIT.java:250]: done https://github.com/dotCMS/core/pull/24034/commits/cb564daa0c443e886c8931149f0ef0a62935fc0f#diff-25dea0f580007d55e9704544452fcf0833029a2294fb814f4dddfb1bb2e4ba9bR249

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPI.java (hunk 6 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPI.java:160]: done https://github.com/dotCMS/core/pull/24034/commits/cb564daa0c443e886c8931149f0ef0a62935fc0f#diff-8bf1865184fd59c5c64882df72bd0c71a403a544777830acf250bbf54ea84d90R160

File: dotCMS/src/main/java/com/dotcms/experiments/business/result/ExperimentResultQueryFactory.java (hunk 7 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/result/ExperimentResultQueryFactory.java:61]: done https://github.com/dotCMS/core/pull/24034/commits/cb564daa0c443e886c8931149f0ef0a62935fc0f#diff-d08b04f8182ef2c2db20eb4d306fa00b9187a0040f09ee21a9c6c1c5227bb754R60

File: dotCMS/src/main/java/com/dotcms/experiments/business/result/MetricExperimentAnalyzer.java (hunk 8 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/result/MetricExperimentAnalyzer.java:7]: done https://github.com/dotCMS/core/pull/24034/commits/cb564daa0c443e886c8931149f0ef0a62935fc0f

**Pull request merged**